### DISCUSSION TITLE
Execution order of functions in UserObject and Material systems

### DISCUSSION THREAD

**Discussion by NorthMagic** [idaholab/moose#26038] [Q&A General]
I want to update material properties in each linear iteration of a PJFNK calculation. So a material object derived from `GenericConstantMaterial` is constructed, and the `computeQpProperties()` is dependent on a self-defined function's return value to update the material properties. The self-defined function is constructed in a UserObject object which is derived from `ElementIntegralUserObject`. What I want is running `execute()` method along all elements and gathering values in `finalize()` method in each linear iteration, then the self-defined function will return a correct value when it is called by `computeQpProperties()`. However, I found that the self-defined function was called before `finalize()` so I got a wrong return value checked in the debug mode.

Here is the related execution order:
```
      0 Linear |R| = 4.176985e+02
[DBG] Computing elemental user objects on LINEAR
[DBG] Ordering of User Objects on block 1
[DBG] Executing ElementUserObject on LINEAR
[DBG] Order of execution:
[DBG] sph
                             
[DBG] Only user objects active on local element/sides are executed
[DBG] Ordering of User Objects on block 2
[DBG] Executing ElementUserObject on LINEAR
[DBG] Order of execution:
[DBG] sph
```
Here is the related part in the input file:
```
[Materials]
  [openmc_mgxs]
    type = OpenMCXSMaterial
    indexed_by_subdomain_name = true
    openmc_userobject = openmc
    sph_userobject = sph
  []
[]

[UserObjects]
  [openmc]
    type = RunOpenMCUO
    mgxs_cell = 'fuel moderator'
   []
   [sph]
      type = SPHFactorUO
      openmc_userobject = openmc
      group_fluxes = 'group_1 group_2'
      execute_on = LINEAR
    []
[]

[Executioner]
  type = Steady
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]
```

**Comment by lindsayad** [comment]
Before executing user objects, materials will be computed. See the relevant source code here:
```c++
void
ComputeUserObjectsThread::onElement(const Elem * elem)
{
  _fe_problem.prepare(elem, _tid);
  _fe_problem.reinitElem(elem, _tid);

  // Set up Sentinel class so that, even if reinitMaterials() throws, we                                              
  // still remember to swap back during stack unwinding.                                                              
  SwapBackSentinel sentinel(_fe_problem, &FEProblem::swapBackMaterials, _tid);
  _fe_problem.reinitMaterials(_subdomain, _tid);

  for (const auto & uo : _element_objs)
  {
    uo->execute();
```

**Reply by GiudGiud** [reply]
When you copied GenericConstantMaterial to make your material, did you make sure to delete this line:

  params.set<MooseEnum>("constant_on") = "SUBDOMAIN";
or is the discretization of your SPH factors by subdomains?
